presto-jmx/src/main/java/com/facebook/presto/connector/jmx/JmxHistoryDumper.java

```diff
@@ -0,0 +1,106 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public class JmxHistoryDumper
+        implements Runnable
+{
+    private static final Logger log = Logger.get(JmxHistoryDumper.class);
+
+    private final JmxHistoricalData jmxHistoricalData;
+    private final JmxMetadata jmxMetadata;
+    private final JmxRecordSetProvider jmxRecordSetProvider;
+    private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor();
+    private final long period;
+    private long lastDumpTimestamp;
+
+    @Inject
+    public JmxHistoryDumper(
+            JmxHistoricalData jmxHistoricalData,
+            JmxMetadata jmxMetadata,
+            JmxRecordSetProvider jmxRecordSetProvider,
+            JmxConnectorConfig jmxConfig)
+    {
+        this.jmxHistoricalData = requireNonNull(jmxHistoricalData, "jmxStatsHolder is null");
+        this.jmxMetadata = requireNonNull(jmxMetadata, "jmxMetadata is null");
+        this.jmxRecordSetProvider = requireNonNull(jmxRecordSetProvider, "jmxRecordSetProvider is null");
+        this.period = requireNonNull(jmxConfig, "jmxConfig is null").getDumpPeriod().toMillis();
+    }
+
+    public void start()
+    {
+        long nowMillis = getNowMillis();
+        long initialDelay = nowMillis - roundToPeriod(nowMillis + period);
+        executor.scheduleAtFixedRate(this, initialDelay, period, MILLISECONDS);
+    }
+
+    private long roundToPeriod(long millis)
+    {
+        return ((millis + period / 2) / period) * period;
+    }
+
+    @Override
+    public void run()
+    {
+        // we are using rounded up timestamp, so that records from different nodes and different
+        // tables will have matching timestamps (for joining/grouping etc)
+        long dumpTimestamp = roundToPeriod(getNowMillis());
+
+        // if something has lagged and next dump has the same timestamp as this one, ignore this dump
+        if (dumpTimestamp == lastDumpTimestamp) {
+            return;
+        }
+        lastDumpTimestamp = dumpTimestamp;
+
+        for (String tableName : jmxHistoricalData.getTables()) {
+            try {
+            JmxTableHandle tableHandle = requireNonNull(
+                    jmxMetadata.getTableHandle(new SchemaTableName(JmxMetadata.HISTORY_SCHEMA_NAME, tableName)),
+                    format("tableHandle is null for table [%s]", tableName));
+
+                List<Object> row = jmxRecordSetProvider.getLiveRow(
+                        tableHandle,
+                        jmxRecordSetProvider.getColumnTypes(tableHandle.getColumnHandles()),
+                        dumpTimestamp);
+                jmxHistoricalData.addRow(tableName, row);
+            }
+            catch (Throwable ex) {
+                log.error(ex, "Error starting JmxHistoryDumper");
```

Comment:
This error message is not accurate, since it could technically be thrown  this class has been running successfully for a while. It'd be better to get the table handles up front (i.e., in the `start()` method)